src/wallets/injected/MathWallet.ts

```diff
@@ -0,0 +1,164 @@
+import isMobile from "is-mobile";
+
+import { mathWalletIcon } from "../icons";
+import { InjectedWallet, WalletModule } from "../Wallet";
+import { transformActions } from "../actions";
+import { InjectedMathWallet } from "./InjectedMathWallet";
+import { transactions, utils } from "near-api-js";
+
+declare global {
+  interface Window {
+    nearWalletApi: InjectedMathWallet | undefined;
+  }
+}
+
+function setupMathWallet(): WalletModule<InjectedWallet> {
+  return function MathWallet({
+    options,
+    provider,
+    emitter,
+    logger,
+    updateState,
+  }) {
+    let wallet: InjectedMathWallet;
+
+    const isInstalled = () => {
+      return !!window.nearWalletApi;
+    };
+
+    const timeout = (ms: number) => {
+      return new Promise((resolve) => setTimeout(resolve, ms));
+    };
+
+    return {
+      id: "math-wallet",
+      type: "injected",
+      name: "Math Wallet",
+      description: null,
+      iconUrl: mathWalletIcon,
+
+      isAvailable() {
+        if (!isInstalled()) {
+          return false;
+        }
+
+        if (isMobile()) {
+          return false;
+        }
+
+        return true;
+      },
+
+      async init() {
+        await timeout(200);
```

Comment:
We need this timeout because if we are logged this method throws an error (even though the extension is installed)  if we don't add a delay :
**MathWallet.ts line 55-57**

```ts
if (!isInstalled()) {
  throw new Error("Wallet not installed");
}
```

I guess this means we are checking if the wallet is installed a bit too soon I will have a look if I can find another way to fix this.

Reply:
Ah I see! Maybe we'll want to consider a more robust solution that polls frequently to reduce the startup time.

Alternatively, maybe the [load](https://developer.mozilla.org/en-US/docs/Web/API/Window/load_event) event could be useful but still could be too soon. It just makes me a little uneasy that we're waiting fixed amount of time. Not sure if extensions have a standardised way of emitting when they're "ready"?

Reply:
Here's an example of a `waitFor` helper (written in JS so I could test with Node's REPL so will need types adding):
```js
const wait = (ms) => {
  return new Promise((resolve) => setTimeout(resolve, ms));
};

const poll = (cb, interval, remaining) => {
  const result = cb();

  if (result) {
    return result;
  }

  if (!remaining) {
    return;
  }

  return wait(interval)
    .then(() => poll(cb, interval, remaining - 1));
}

const waitFor = async (cb, opts = {}) => {
  const { timeout = 200, interval = 50 } = opts;

  return Promise.race([
    wait(timeout)
      .then(() => {
        throw new Error("Exceeded timeout")
      }),
    poll(cb, interval, Math.floor(timeout / interval))
  ])
}

// Example
(async () => {
  let x = false;

  setTimeout(() => x = true, 500);

  try {
    const res = await waitFor(() => x, { timeout: 550, interval: 50 });

    console.log("res:", res);
  } catch (err) {
    console.log("err:", err.message);
  }
})();
```